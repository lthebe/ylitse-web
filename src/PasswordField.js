import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { FormControl, FormHelperText } from 'material-ui/Form';
import IconButton from 'material-ui/IconButton';
import Input, { InputLabel, InputAdornment } from 'material-ui/Input';
import { withStyles } from 'material-ui/styles';
import VisibilityIcon from 'material-ui-icons/Visibility';
import VisibilityOffIcon from 'material-ui-icons/VisibilityOff';

const styles = () => ({
    iconButton: {
        marginLeft: 20,
        marginRight: -12,
    },
});

class PasswordField extends Component {
    static propTypes = {
        classes: PropTypes.shape({ iconButton: PropTypes.string }).isRequired,
        label: PropTypes.string,
        helperText: PropTypes.string,
        required: PropTypes.bool,
        error: PropTypes.bool,
        className: PropTypes.string,
    }

    static defaultProps = {
        label: 'Password',
        helperText: '',
        required: false,
        error: false,
        className: '',
    }

    constructor(props) {
        super(props);

        this.state = { showPassword: false };
    }

    toggleVisibility = () => {
        this.setState({ showPassword: !this.state.showPassword });
    };

    render() {
        const {
            label, helperText, required, error, className, classes, ...props
        } = this.props;
        const { showPassword } = this.state;

        return (
            <FormControl required={required} className={className}>
                <InputLabel error={error}>{label}</InputLabel>
                <Input
                    {...props}
                    type={showPassword ? 'text' : 'password'}
                    endAdornment={
                        <InputAdornment position="end">
                            <IconButton
                                className={classes.iconButton}
                                onClick={this.toggleVisibility}
                            >
                                {showPassword ?
                                    <VisibilityOffIcon /> :
                                    <VisibilityIcon />
                                }
                            </IconButton>
                        </InputAdornment>
                    }
                />
                {helperText &&
                    <FormHelperText error={error}>{helperText}</FormHelperText>
                }
            </FormControl>
        );
    }
}

export default withStyles(styles)(PasswordField);
