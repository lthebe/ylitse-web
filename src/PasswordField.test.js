import React from 'react';
import { shallow, mount } from 'enzyme';
import PasswordField from './PasswordField';

describe('PasswordField', () => {
    test('renders correctly', () => {
        // PasswordField is a HOC, so diving
        const wrapper = shallow(<PasswordField />).dive();

        expect(wrapper).toMatchSnapshot();
    });

    test('click icon toggles the visibility', () => {
        const wrapper = mount(<PasswordField />);

        expect(wrapper.find('Input').prop('type')).toBe('password');
        wrapper.find('IconButton').simulate('click');
        expect(wrapper.find('Input').prop('type')).toBe('text');
    });

    test('helperText displayed in FormHelperText', () => {
        const wrapper = mount(<PasswordField />);
        wrapper.setProps({ helperText: 'Not good password' });

        expect(wrapper.find('FormHelperText').length).toBe(1);
    });
});
