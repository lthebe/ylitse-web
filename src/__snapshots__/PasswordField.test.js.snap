// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordField renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PasswordField
    className=""
    classes={
      Object {
        "iconButton": "PasswordField-iconButton-1",
      }
    }
    error={false}
    helperText=""
    label="Password"
    required={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <WithStyles(InputLabel)
          error={false}
        >
          Password
        </WithStyles(InputLabel)>,
        <WithStyles(Input)
          endAdornment={
            <WithStyles(InputAdornment)
              position="end"
            >
              <WithStyles(IconButton)
                className="PasswordField-iconButton-1"
                onClick={[Function]}
              >
                <pure(Visibility) />
              </WithStyles(IconButton)>
            </WithStyles(InputAdornment)>
          }
          type="password"
        />,
        "",
      ],
      "className": "",
      "required": false,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Password",
          "error": false,
        },
        "ref": null,
        "rendered": "Password",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "endAdornment": <WithStyles(InputAdornment)
            position="end"
          >
            <WithStyles(IconButton)
              className="PasswordField-iconButton-1"
              onClick={[Function]}
            >
              <pure(Visibility) />
            </WithStyles(IconButton)>
          </WithStyles(InputAdornment)>,
          "type": "password",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "",
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <WithStyles(InputLabel)
            error={false}
          >
            Password
          </WithStyles(InputLabel)>,
          <WithStyles(Input)
            endAdornment={
              <WithStyles(InputAdornment)
                position="end"
              >
                <WithStyles(IconButton)
                  className="PasswordField-iconButton-1"
                  onClick={[Function]}
                >
                  <pure(Visibility) />
                </WithStyles(IconButton)>
              </WithStyles(InputAdornment)>
            }
            type="password"
          />,
          "",
        ],
        "className": "",
        "required": false,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Password",
            "error": false,
          },
          "ref": null,
          "rendered": "Password",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "endAdornment": <WithStyles(InputAdornment)
              position="end"
            >
              <WithStyles(IconButton)
                className="PasswordField-iconButton-1"
                onClick={[Function]}
              >
                <pure(Visibility) />
              </WithStyles(IconButton)>
            </WithStyles(InputAdornment)>,
            "type": "password",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "",
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
